core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala

```diff
@@ -0,0 +1,746 @@
+        migration.newGoogleProjectId.getOrElse(throw noGoogleProjectError(migration))
+      })
+
+      workspace <- getWorkspace(migration.workspaceId)
+      tmpBucketName <- MigrateAction.liftIO(randomSuffix("terra-workspace-migration-"))
+      _ <- createBucketInSameRegion(
+        migration,
+        GcsBucketName(workspace.bucketName),
+        googleProjectId,
+        GcsBucketName(tmpBucketName)
+      )
+
+      created <- nowTimestamp
+      _ <- inTransaction { _ =>
+        workspaceMigrations
+          .filter(_.id === migration.id)
+          .map(r => (r.tmpBucket, r.tmpBucketCreated))
+          .update((tmpBucketName.some, created.some))
+      }
+    } yield ()
+
+
+  final def issueWorkspaceBucketTransferJob: MigrateAction[Unit] =
+    for {
+      migration <- findMigration { m =>
+        m.tmpBucketCreated.isDefined && m.workspaceBucketTransferJobIssued.isEmpty
+      }
+
+      tmpBucketName <- MigrateAction.liftIO(IO {
+        migration.tmpBucketName.getOrElse(throw noGoogleProjectError(migration))
+      })
+
+      workspace <- getWorkspace(migration.workspaceId)
+      _ <- startBucketTransferJob(
+        migration.id,
+        GcsBucketName(workspace.bucketName),
+        tmpBucketName
+      )
+
+      issued <- nowTimestamp
+      _ <- inTransaction { _ =>
+        workspaceMigrations
+          .filter(_.id === migration.id)
+          .map(_.workspaceBucketTransferJobIssued)
+          .update(issued.some)
+      }
+    } yield ()
+
+
+  final def deleteWorkspaceBucket: MigrateAction[Unit] =
+    for {
+      migration <- findMigration { m =>
+        m.workspaceBucketTransferred.isDefined && m.workspaceBucketDeleted.isEmpty
+      }
+
+      workspace <- getWorkspace(migration.workspaceId)
+      storageService <- MigrateAction.asks(_.storageService)
+      _ <- MigrateAction.liftIO {
+        for {
+          successOpt <- storageService.deleteBucket(
+            GoogleProject(workspace.googleProjectId.value),
+            GcsBucketName(workspace.bucketName),
+            isRecursive = true
+          ).compile.last
+
+          _ <- IO.raiseUnless(successOpt.contains(true)) {
+            noWorkspaceBucketError(migration, GcsBucketName(workspace.bucketName))
+          }
+        } yield ()
+      }
+
+      deleted <- nowTimestamp
+      _ <- inTransaction { _ =>
+        workspaceMigrations
+          .filter(_.id === migration.id)
+          .map(_.workspaceBucketDeleted)
+          .update(deleted.some)
+      }
+    } yield ()
+
+
+  final def createFinalWorkspaceBucket: MigrateAction[Unit] =
+    for {
+      migration <- findMigration { m =>
+        m.workspaceBucketDeleted.isDefined && m.finalBucketCreated.isEmpty
+      }
+
+      (googleProjectId, tmpBucketName) <- MigrateAction.liftIO(IO {
+        (
+          migration.newGoogleProjectId.getOrElse(throw noGoogleProjectError(migration)),
+          migration.tmpBucketName.getOrElse(throw noTmpBucketError(migration))
+        )
+      })
```

Comment:
this block is duplicated in `deleteTemporaryBucket`(). is it worth refactoring?

Reply:
Good idea